### PR TITLE
Initial PR for internal storage API

### LINKED ISSUE #18885: [RFC] Common Ray persistent storage config
We have a number of different components and libraries that have similar persistent storage client interfaces:

- Ray Serve (for checkpointing)
- Ray Tune (for checkpointing)
- Ray Workflows (for persisting code & workflow outputs)
- Object spilling (for spilling objects to remote storage)
- ...

In order to minimize code duplication and provide building blocks for library developers, we propose adding a common storage client abstraction to Ray. This will be a generic KV-store interface that all of the libraries can use to interact with storage. The backend for the storage can either be configured at the cluster level (by passing a storage path to `ray start --head`) or manually at the library level by calling the storage client constructors directly.

Interface (WIP):
```python
def get_storage_client(namespace: Optional[str] = None) -> StorageClient:
    """Namespace can be used to avoid key collisions between libraries. Could also respect Ray namespaces."""
    pass

class StorageClient(abc.ABC):
    def put(self, key: str, val: bytes, overwrite=False):
        pass

    def get(self, key: str) -> bytes:
        pass

    def exists(self, key: str) -> bool:
        pass

    def delete(self, key: str) -> bool:
        pass

    def size(self, key: str) -> int:
        pass

    def scan_prefix(self, key_prefix: str) -> List[str]: # Necessary for Ray Workflows.
        pass
```

Configuring the storage client at the cluster level. We could lean on [smart_open](https://pypi.org/project/smart-open/) for this.
```
ray start --head --storage-path=s3://path/path
ray start --head --storage-path=azure://my_bucket/my_blob
```

Libraries would use the client by calling `ray.storage.get_storage_client()`:
```python
client = ray.storage.get_storage_client()
client.put("hello", b"world")
assert client.get("hello") == b"world"
```

This could trivially replace the Serve [KV store interface](https://github.com/ray-project/ray/blob/master/python/ray/serve/storage/kv_store.py). If people agree on the high-level proposal & basic interface, next step will be to do an audit on other components' interfaces to find the minimal API we can use to unify them.

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

This adds the initial PR for a common internal storage API. In a follow-up PR, documentation and object spilling support will be added.

TODO:
- [x] Remove object / checkpointing support from initial version
- [x] Unit tests for main impl
- [x] Custom storage modules
- [x] Fix directory traversal via ..
- [x] Rename namespace to prefix
- [ ] Spilling impl
- [ ] Port unit tests for spilling impl

## Related issue number

Closes https://github.com/ray-project/ray/issues/18885

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/internal/storage.py [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+            Whether the blob was deleted.
+        """
+        full_path = self._resolve_path(path)
+        try:
+            self.fs.delete_file(full_path)
+            return True
+        except FileNotFoundError:
+            return False
+
+    def get_info(self, path: str) -> "pyarrow.fs.FileInfo":
```
Comment (suquark) [inline python/ray/internal/storage.py:155]:
```suggestion
    def get_info(self, path: str) -> "Optional[pyarrow.fs.FileInfo]":
```

File: python/ray/node.py [hunk 2]
```diff
@@ -21,6 +21,7 @@
 
 import ray
 import ray.ray_constants as ray_constants
+import ray.internal.storage as storage
```
Comment (suquark) [inline python/ray/node.py:24]: `from ray.internal import storage`?

File: python/ray/workers/default_worker.py [hunk 3]
```diff
@@ -163,6 +170,7 @@
     if mode == ray.RESTORE_WORKER_MODE or mode == ray.SPILL_WORKER_MODE:
         from ray import external_storage
 
+        ray.storage.impl._init_storage(args.storage, is_head=False)
```
Comment (suquark) [inline python/ray/workers/default_worker.py:173]: What's the difference between `ray.storage.impl._init_storage` and `ray.internal.storage._init_storage`? Why use `ray.storage.impl._init_storage`  here?

Comment (ericl) [reply python/ray/workers/default_worker.py:173]: Fixed.

**Review by suquark — APPROVED** [review]
LGTM

**Pull request merged**